File: app/src/main/java/com/juniperphoton/myerlist/api/CloudService.java
package com.juniperphoton.myerlist.api;

import com.juniperphoton.myerlist.api.response.AddToDoResponse;
import com.juniperphoton.myerlist.api.response.CateResponse;
import com.juniperphoton.myerlist.api.response.CheckUserResponse;
import com.juniperphoton.myerlist.api.response.CommonResponse;
import com.juniperphoton.myerlist.api.response.GetOrderResponse;
import com.juniperphoton.myerlist.api.response.GetSaltResponse;
import com.juniperphoton.myerlist.api.response.LoginResponse;
import com.juniperphoton.myerlist.api.response.GetToDosResponse;
import com.juniperphoton.myerlist.api.response.RegisterResponse;
import com.juniperphoton.myerlist.util.AppConfig;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import rx.Observable;

@SuppressWarnings("UnusedDeclaration")
public class CloudService {
    public static final String BASE_URL = "http://juniperphoton.net/schedule/";
    private static final int DEFAULT_TIMEOUT = 10;

    private Retrofit mRetrofit;

    public CloudService() {
        OkHttpClient.Builder builder = new OkHttpClient.Builder();
        builder.connectTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS);

        mRetrofit = new Retrofit.Builder()
                .client(builder.build())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(BASE_URL)
                .build();
    }

    private static class SingletonHolder {
        private static final CloudService INSTANCE = new CloudService();
    }

    public static CloudService getInstance() {
        return SingletonHolder.INSTANCE;
    }

    public Observable<CheckUserResponse> checkUserExist(String email) {
        return mRetrofit.create(UserService.class).checkUserExist(email);
    }

    public Observable<GetSaltResponse> getSalt(String email) {
        return mRetrofit.create(UserService.class).getSalt(email);
    }

    public Observable<LoginResponse> login(String email, String pwd) {
        return mRetrofit.create(UserService.class).login(email, pwd);
    }

    public Observable<RegisterResponse> register(String email, String pwd) {
        return mRetrofit.create(UserService.class).register(email, pwd);
    }

    public Observable<CateResponse> getCategories() {
        return mRetrofit.create(CategoryService.class).getCategory(AppConfig.getSid(), AppConfig.getAccessToken());
    }

    public Observable<GetToDosResponse> getToDos() {
        return mRetrofit.create(ToDoService.class).getToDos(AppConfig.getSid(), AppConfig.getAccessToken(), AppConfig.getSid());
    }

    public Observable<GetOrderResponse> getOrders() {
        return mRetrofit.create(ToDoService.class).getOrders(AppConfig.getSid(), AppConfig.getAccessToken(), AppConfig.getSid());
    }

    public Observable<CommonResponse> setOrders(String order) {
        return mRetrofit.create(ToDoService.class).setOrders(AppConfig.getSid(), AppConfig.getAccessToken(), AppConfig.getSid(), order);
    }

    public Observable<CommonResponse> setIsDone(String id, String isDone) {
        return mRetrofit.create(ToDoService.class).setDone(AppConfig.getSid(), AppConfig.getAccessToken(), id, isDone);
    }

    public Observable<CommonResponse> deleteToDo(String id) {
        return mRetrofit.create(ToDoService.class).deleteToDo(AppConfig.getSid(), AppConfig.getAccessToken(), id);
    }

    public Observable<AddToDoResponse> addToDo(String time, String content, String isDone, String cate) {
        return mRetrofit.create(ToDoService.class).addToDo(AppConfig.getSid(), AppConfig.getAccessToken(),
                AppConfig.getSid(), time, content, isDone, cate);
    }

    public Observable<CommonResponse> updateToDo(String id, String time, String content, String cate) {
        return mRetrofit.create(ToDoService.class).updateToDo(AppConfig.getSid(), AppConfig.getAccessToken(),
                id, time, content, cate);
    }

    public Observable<CommonResponse> updateToDoCategories(String str) {
        return mRetrofit.create(CategoryService.class).setCategory(AppConfig.getSid(), AppConfig.getAccessToken(),
                str, AppConfig.getSid(), AppConfig.getAccessToken());
    }
}

File: app/src/main/java/com/juniperphoton/myerlist/api/CloudService.kt
package com.juniperphoton.myerlist.api

import com.juniperphoton.myerlist.api.response.*
import com.juniperphoton.myerlist.util.AppConfig
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import rx.Observable
import java.util.concurrent.TimeUnit

object CloudService {
    private val BASE_URL = "http://juniperphoton.net/schedule/"
    private val DEFAULT_TIMEOUT = 10

    private val retrofit: Retrofit

    init {
        val builder = OkHttpClient.Builder()
        builder.connectTimeout(DEFAULT_TIMEOUT.toLong(), TimeUnit.SECONDS)

        retrofit = Retrofit.Builder()
                .client(builder.build())
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .baseUrl(BASE_URL)
                .build()
    }

    fun checkUserExist(email: String): Observable<CheckUserResponse> {
        return retrofit.create(UserService::class.java).checkUserExist(email)
    }

    fun getSalt(email: String): Observable<GetSaltResponse> {
        return retrofit.create(UserService::class.java).getSalt(email)
    }

    fun login(email: String, pwd: String): Observable<LoginResponse> {
        return retrofit.create(UserService::class.java).login(email, pwd)
    }

    fun register(email: String, pwd: String): Observable<RegisterResponse> {
        return retrofit.create(UserService::class.java).register(email, pwd)
    }

    fun getCategories(): Observable<CateResponse> = retrofit.create(CategoryService::class.java)
            .getCategory(AppConfig.sid!!, AppConfig.accessToken!!)

    fun getToDos(): Observable<GetToDosResponse> = retrofit.create(ToDoService::class.java)
            .getToDos(AppConfig.sid!!, AppConfig.accessToken!!, AppConfig.sid)

    fun getOrders(): Observable<GetOrderResponse> = retrofit.create(ToDoService::class.java)
            .getOrders(AppConfig.sid!!, AppConfig.accessToken!!, AppConfig.sid)

    fun setOrders(order: String): Observable<CommonResponse> {
        return retrofit.create(ToDoService::class.java)
                .setOrders(AppConfig.sid!!, AppConfig.accessToken!!, AppConfig.sid, order)
    }

    fun setIsDone(id: String, isDone: String): Observable<CommonResponse> {
        return retrofit.create(ToDoService::class.java)
                .setDone(AppConfig.sid!!, AppConfig.accessToken!!, id, isDone)
    }

    fun deleteToDo(id: String): Observable<CommonResponse> {
        return retrofit.create(ToDoService::class.java).deleteToDo(AppConfig.sid!!, AppConfig.accessToken!!, id)
    }

    fun addToDo(time: String, content: String, isDone: String, cate: String): Observable<AddToDoResponse> {
        return retrofit.create(ToDoService::class.java).addToDo(AppConfig.sid!!, AppConfig.accessToken!!,
                AppConfig.sid, time, content, isDone, cate)
    }

    fun updateToDo(id: String, time: String, content: String, cate: String): Observable<CommonResponse> {
        return retrofit.create(ToDoService::class.java).updateToDo(AppConfig.sid!!, AppConfig.accessToken!!,
                id, time, content, cate)
    }

    fun updateToDoCategories(str: String): Observable<CommonResponse> {
        return retrofit.create(CategoryService::class.java).setCategory(AppConfig.sid!!, AppConfig.accessToken!!,
                str, AppConfig.sid, AppConfig.accessToken)
    }
}